<a href="https://colab.research.google.com/github/rileygoetz/DS1002-S24/blob/main/Assignments/14_TextGeneration.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Final homework assignment:  generating text from a corpus

Due Friday 3 May at 5 pm

We begin by loading all of Shakespeare's plays into a text file, and separating them into individual words.  This is made easy by the fact that someone already made a giant text file of these plays for us, and put it online.

[Note that by changing the url in the following code chunk, you can load any text file that lives online in this way . . . see below.]

Run the following code.  It should output a list of ten words, starting with 'From' and ending with 'rose'.

In [1]:
from urllib.request import urlopen            #package for reading info from the web
url='https://ocw.mit.edu/ans7870/6/6.006/s08/lecturenotes/files/t8.shakespeare.txt'   #web address
text = urlopen(url).read().decode('utf-8')    #open web address, read file, convert to text
shakespeare_words=text.split()[1651:]         #split text into words (not including preamble)
shakespeare_words[:10]                        #display the first 10 words

['From',
 'fairest',
 'creatures',
 'we',
 'desire',
 'increase,',
 'That',
 'thereby',
 "beauty's",
 'rose']

## 1. `nextwords` practice

Create a practice function called `createNextwordsPractice` which takes a list of words called `wordlist` and returns a dictionary called `nextwordsPractice`. The keys of `nextwordsPractice` should be all the words in `wordlist`, and the value for any given key should be a list of all the words that follow the key word somewhere in the original `wordlist`.



In [2]:
def createNextwordsPractice(wordlist):
  nextwordsPractice = {}
  for i in range(len(wordlist) - 1):
      if wordlist[i] not in nextwordsPractice:
            nextwordsPractice[wordlist[i]] = []
      nextwordsPractice[wordlist[i]].append(wordlist[i + 1])
  return nextwordsPractice

Run the following code:

In [3]:
practice = createNextwordsPractice(shakespeare_words)
practice['From']

['fairest',
 'his',
 'sullen',
 'limits',
 'hands',
 'whence',
 'where',
 'me',
 'this',
 'thy',
 'hence',
 'thee,',
 'you',
 'heaven',
 'her',
 'simple',
 'whence',
 'below',
 'lowest',
 'courtly',
 'son',
 'the',
 'the',
 'Sicyon,',
 'Alexandria',
 'Silvius,',
 'mine',
 'the',
 'firm',
 'Egypt',
 'Antony',
 'thine',
 'that',
 'my',
 'me',
 'henceforth',
 'tyrant',
 'seventeen',
 'miserable',
 'whom',
 'my',
 'whence,',
 'me',
 'me',
 'slaves',
 'every',
 'rascals',
 'whom',
 'where',
 "th'",
 'Rome',
 'off',
 'these',
 'twelve',
 "th'",
 'whence?',
 'Rome.',
 'thy',
 'Gallia',
 'fairies',
 'off',
 "th'",
 'most',
 'thy',
 'youth',
 'every',
 'whose',
 'so',
 'one',
 'this',
 'the',
 'east',
 'action',
 'my',
 'this',
 'Leonati',
 'her',
 'chance',
 'south',
 'our',
 'whence',
 'the',
 'top',
 'this',
 'our',
 'that',
 'me,',
 'fashion',
 'the',
 'another',
 'Hamlet?',
 'whence',
 'whom',
 "Rumour's",
 'whence',
 'the',
 'a',
 'a',
 'enemies',
 'every',
 'this',
 'the',
 'open',
 'who

In [4]:
practice['rose']

['might',
 'looks',
 'is',
 'of',
 'Of',
 'may',
 'will',
 'of',
 'From',
 'and',
 'of',
 'both',
 'with',
 'from',
 'with',
 'with',
 'side.',
 'red,',
 'too.',
 'in',
 'a',
 'a',
 'I',
 'against',
 'that',
 'and',
 'from',
 'again,',
 'Lest',
 'against',
 'Than',
 "distill'd",
 'on',
 'up',
 'in',
 'wither.',
 'and',
 'By',
 'again?',
 'he']

In [5]:
practice['rose.']

['It']

## 2. `nextwords` for real:  new sentences.

Next, write a new function called `createNextwords` which takes a `wordlist` and returns a dictionary of `nextwords`, but now with a new additional feature.  
In addition to the keys from problem 1, add a special key to the dictionary, which will be the symbol '%'.  This key will signify the first word of a sentence. When scanning the wordlist to construct the `nextwords` dictionary, normally you would add `wordlist[i+1]` to the list `nextwords[wordlist[i]]`. However, if `wordlist[i]` ends with a period, then you should instead add `wordlist[i+1]` to the list nextwords['%']. Also add the very first word of `wordlist` to nextwords['%'].


You should be able to produce this output:
```
> shakespeare_nextwords=createNextwords(shakespeare_words)
> shakespeare_nextwords['Rose']
['at', 'or', 'or', 'and']

> shakespeare_nextwords['rose.']
[]

> len(shakespeare_nextwords['%'])
76543
```

In [6]:
def createNextwords(wordlist):
    nextwords = {}
    nextwords['%'] = [wordlist[0]]
    for i in range(len(wordlist) - 1):
        current_word = wordlist[i]
        next_word = wordlist[i + 1]
        if current_word.endswith('.'):
            if '%' not in nextwords:
                nextwords['%'] = []
            nextwords['%'].append(next_word)
        else:
            if current_word not in nextwords:
                nextwords[current_word] = []
            nextwords[current_word].append(next_word)

    return nextwords

Run the following code:

In [7]:
shakespeare_nextwords = createNextwords(shakespeare_words)

print(shakespeare_nextwords['Rose'])
print(len(shakespeare_nextwords['%']))

['at', 'or', 'or', 'and']
76543


In [8]:
print(len(shakespeare_nextwords['rose']))

40


In [16]:
print(shakespeare_nextwords['rose.'])

KeyError: 'rose.'

I tried to figure out why this code wouldn't run but I was having a very hard time.

## 3.  Make your own Shakespeare play

Okay, that may be a bit of an exaggeration.  Anyway, next up is text generation.  
We'll need the choice function:

In [10]:
from random import choice
import random

This function chooses a random element from a list.  Run this code a few times:

In [11]:
choice([1, 17, 34, 289])

1

Now, create a function called `makePlay(nextwords,n)` which takes a dictionary `nextwords` and a number `n`, and follows these steps:
1. Define an empty `newplay` string.  (I did this for you.)
2. Set `currentword` equal to '%'.
3. Start a for-loop that will run `n` times.
4. Randomly choose a word from the list `nextwords[currentword]`.
5. Add the chosen word to `newplay`, followed by a space `' '` character.
6. If the chosen word ends in a period, add a newline character '\n' and set `currentword` equal to `%`.
7. Otherwise, set `currentword` equal to the chosen word.
8. Go back to step 4.
9. When all `n` iterations have finished, your function should return `newplay`.

In [12]:
def makePlay(nextwords,n):
  newplay=''
  currentword = '%'
  for _ in range(n):
        if currentword not in nextwords or not nextwords[currentword]:
            currentword = '%'
        chosen_word = random.choice(nextwords[currentword])
        newplay += chosen_word + ' '
        if chosen_word.endswith('.'):
            newplay += '\n'
            currentword = '%'
        else:
            currentword = chosen_word
  return newplay

Now create a 1000 word Shakesperian play by running the following code:

In [17]:
nextwords=createNextwords(shakespeare_words)
newplay=makePlay(nextwords,1000)
print(newplay)

The king-becoming graces, As guilty of reck'ning. 
FIRST MURTHERER. 
SECOND LORD. 
Friar. 
Yes, Warwick, I bar my all too fond, but a bitter tongue! I am thy sweet lords, I marvel why Stands not strike thee well; what's more Of mighty suit; Which renders you- is this? IAGO. 
Rom. 
Pedro. 
Why, this and a copy-book. 
Yes, as a hill Of what else conclude that would not Barbason: you must wed Demetrius. 
Item, Sack and be blown up his manners. 
What an instant came sighing breast Teems and others; our striving to lament or the beetle brows are my varlet; I'll show thee, Nero, Play on her thus. 
There is the latter times. 
The palace. 
Touching the matter? CLEOPATRA. 
Orsino, youth. 
HASTINGS. 
A maid, they told thee into you, is a box on our minister, Throw over the gall. 
I am content; A street. 
TITUS. 
Our holy rood, She hath outrun on thy cheek. 
That if you incens'd against him as strange infirmity, for thy rose will steal at last, the Earl of my countryman- His pow'rful sun, His gar

At this point you might enjoy finding other sources of text on the internet and replacing the url at the top of this notebook. You can use song lyrics, news articles, etc. You just need to find files ending with '.txt'. Alternatively, you can copy and paste any block of text from any file into colab, assign it to the variable `text`, and then use wordlist=text.split() to get a list of words.

## Bonus:

Recall the following recursive function that we used to define the Fibonacci numbers:

In [14]:
def fibonacci(n):
  # Base cases: F(0) = 0, F(1) = 1
  if n == 0:
    return 0
  elif n == 1:
    return 1
  else:
    # Recursive case: F(n) = F(n-1) + F(n-2)
    return fibonacci(n - 1) + fibonacci(n - 2)

fibonacci(17)

1597

Here is the question.  If you run this line of code:  
`fibonacci(17)`  
how many times does the function `fibonacci()` get called?  
(The answer is not 1597, which is the value returned by the function.)

In [15]:
def fibonacci(n):
    global count  # We use a global variable to keep track of the number of calls
    count += 1
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fibonacci(n - 1) + fibonacci(n - 2)

count = 0
fibonacci(17)
print("fibonacci(17) was called", count, "times.")

fibonacci(17) was called 5167 times.


fibonacci(17) gets called a total of 5167 times.